include/libdnf/rpm/solv_sack.hpp

```diff
@@ -0,0 +1,140 @@
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+
+#ifndef LIBDNF_RPM_SACK_HPP
+#define LIBDNF_RPM_SACK_HPP
+
+#include "libdnf/utils/exception.hpp"
+
+#include <memory>
+
+namespace libdnf {
+
+class Base;
+
+}  // namespace libdnf
+
+namespace libdnf::rpm {
+
+
+class PackageSet;
+
+
+struct PackageId {
+public:
+    PackageId() = default;
+    explicit PackageId(int id);
+
+    bool operator==(const PackageId & other) const noexcept { return id == other.id; };
+    bool operator!=(const PackageId & other) const noexcept { return id != other.id; };
+
+    int id{0};
+};
+
+struct ReldepId {
+public:
+    ReldepId() = default;
+    explicit ReldepId(int id);
+
+    bool operator==(const ReldepId & other) const noexcept { return id == other.id; };
+    bool operator!=(const ReldepId & other) const noexcept { return id != other.id; };
+
+    int id{0};
+};
+
+inline PackageId::PackageId(int id) : id(id) {}
+
+inline ReldepId::ReldepId(int id) : id(id) {}
+
+
+// forward declarations
+class Package;
+class Reldep;
+class ReldepList;
+class Repo;
+
+class SolvSack {
+public:
+    class Exception : public RuntimeError {
+    public:
+        using RuntimeError::RuntimeError;
+        const char * get_domain_name() const noexcept override { return "libdnf::rpm::SolvSack"; }
+        const char * get_name() const noexcept override { return "Exception"; }
+        const char * get_description() const noexcept override { return "rpm::SolvSack exception"; }
+    };
+
+    class NoCapability : public Exception {
+    public:
+        using Exception::Exception;
+        const char * get_domain_name() const noexcept override { return "libdnf::rpm::SolvSack"; }
+        const char * get_name() const noexcept override { return "NoCapability"; }
+        const char * get_description() const noexcept override {
+            return "repository does not provide required metadata type";
+        }
+    };
+
+    // LoadRepoFlags:
+    // @NONE:                    Load only main solvables
+    // @USE_FILELISTS:           Add the filelists metadata
+    // @USE_PRESTO:              Add presto deltas metadata
+    // @USE_UPDATEINFO:          Add updateinfo metadata
+    // @USE_OTHER:               Add other metadata
+    enum class LoadRepoFlags {
+        NONE = 0,
+        USE_FILELISTS = 1 << 1,
+        USE_PRESTO = 1 << 2,
+        USE_UPDATEINFO = 1 << 3,
+        USE_OTHER = 1 << 4,
+    };
+
+    SolvSack(Base & base);
+    ~SolvSack();
+
+    /// Loads rpm::Repo into SolvSack.
+    void load_repo(Repo & repo, bool build_cache, LoadRepoFlags flags);
+
+private:
+    friend Package;
+    friend PackageSet;
+    friend Reldep;
+    friend ReldepList;
+    class Impl;
+    std::unique_ptr<Impl> pImpl;
+};
+
+inline constexpr SolvSack::LoadRepoFlags operator|(SolvSack::LoadRepoFlags lhs, SolvSack::LoadRepoFlags rhs) {
```

Comment:
What about to make a template/macro that will aotomaticaly define operators.

Reply:
In fact I wanted to discuss about it.